### ISSUE TITLE
Possible to modify initial LRP score to compute CLRP(contrastive lrp)?

### ISSUE THREAD

**Issue by 1058433796** [pytorch/captum#1275]
CLRP uses different inital LRP score, and I want to use lrp method to compute CLRP. Is it possible?

**Comment by craymichael** [comment]
Apologies for the long delay here.

Short answer: Captum's current `LRP` API can select a target output, but it cannot directly set an arbitrary initial output relevance vector. That means CLRP type 1 is not supported as a single `LRP.attribute` call, because CLRP type 1 initializes the contrastive relevance from the non-target outputs rather than from one selected scalar target.

For CLRP type 1, the closest workaround is to run LRP once for the target class and once for each non-target class, combine the non-target maps with a sign correction, normalize, and subtract:

```python
from captum.attr import LRP

def normalize_relevance(relevance):
    denom = relevance.flatten(1).sum(dim=1).reshape(
        -1, *([1] * (relevance.dim() - 1))
    )
    return relevance / torch.where(
        denom.abs() > 1e-12, denom, torch.ones_like(denom)
    )

lrp = LRP(model)
logits = model(inputs)  # pre-softmax outputs
target_relevance = lrp.attribute(inputs, target=target)
dual_relevance = torch.zeros_like(target_relevance)

for class_idx in range(logits.shape[1]):
    if class_idx == target:
        continue

    class_relevance = lrp.attribute(inputs, target=class_idx)

    # Captum's per-class LRP relevance sums to that class score. CLRP type 1
    # seeds non-target relevance with abs(output), so negative logits need their
    # relevance sign flipped before accumulating the contrastive map.
    sign = torch.sign(logits[:, class_idx]).reshape(
        -1, *([1] * (class_relevance.dim() - 1))
    )
    dual_relevance = dual_relevance + sign * class_relevance

clrp_relevance = torch.clamp(
    normalize_relevance(target_relevance) - normalize_relevance(dual_relevance),
    min=0,
)
```

This is not as efficient or as clean as native CLRP support, since it needs one LRP run per non-target class. It also relies on the usual linearity of propagated LRP relevance once the model activations and rules are fixed.

For CLRP type 2, a custom rule is closer to sufficient. Type 2 still starts from the target output, so `target=...` is usable. The dual pass changes the final linear layer's relevance propagation, which can be approximated by attaching a custom `PropagationRule` to the final `nn.Linear` layer:

```python
from captum.attr import LRP
from captum.attr._utils.lrp_rules import EpsilonRule

class NegatedFinalLinearRule(EpsilonRule):
    def _manipulate_weights(self, module, inputs, outputs):
        module.weight.data = -module.weight.data
        if module.bias is not None:
            module.bias.data = torch.zeros_like(module.bias.data)

lrp = LRP(model)

target_relevance = lrp.attribute(inputs, target=target)

model.classifier.rule = NegatedFinalLinearRule()  # use your model's final Linear module
dual_relevance = lrp.attribute(inputs, target=target)

clrp_relevance = torch.clamp(
    normalize_relevance(target_relevance) - normalize_relevance(dual_relevance),
    min=0,
)
```

That is still a workaround, not first-class CLRP support. It assumes the classifier head is an actual `nn.Linear` module rather than a functional `F.linear` call, and exact parity with the CLRP reference depends on matching the rest of the LRP rules used in that implementation. We also have PR #1841 open to preserve user-attached custom LRP rules across repeated attribution calls, so custom rules like this should persist after `LRP.attribute` once that change lands.

Since the snippets above should give enough pointers to run CLRP-style attribution using Captum today, I'll close this for now. We can track first-class CLRP support or a general `initial_relevance` / `output_relevance` option separately if there is more demand. That kind of API would let users pass an output-shaped relevance tensor directly, but it would need to generalize the current scalar-target path and update convergence-delta semantics.
